### ISSUE TITLE
Regression (0.22.x) Execution continues after exception

### ISSUE THREAD

**Issue by kenjones-cisco** [yeoman/generator#888]
Under 0.20.x and 0.21.x, after an exception is thrown, the execution ends as expected but as of 0.22.x the generator execution will continue.

Throwing error like the following:
```javascript
this.env.error(new Error('missing required input `apiPath`'));
```

Also tried just simply performing `throw` and the same results.

Test Case:
```javascript
    it('no api file', function (done) {
        var base = testutil.makeBase('spec');

        base.args = ['foo'];
        base.prompt.apiPath = null;

        testutil.runGen(base, function (err) {
            expect(err).toBeDefined();
            expect(err.message).toMatch(/missing required input/);
            done();
        });
    });
```

runGen
```javascript
var helpers = require('yeoman-test');

util.runGen = function (config, done) {
    this.errorCalled = false;

    function error(err) {
        if (this.errorCalled) {
            return;
        }
        this.errorCalled = true;
        done(err);
    }

    function end(err) {
        if (this.errorCalled) {
            return;
        }
        this.errorCalled = true;
        done(err);
    }

    helpers.run(path.join(BASE_DIR, config.gen))
        .withOptions(config.options)
        .withArguments(config.args)
        .withPrompts(config.prompt)
        .on('error', error)
        .on('end', end);
};
```

When running against pre 0.22.x, the test runs successfully, but as soon as I upgrade to 0.22.x, I get additional errors being thrown from addition tasks being queued up and executed by tasks that come after the one throwing the error.

Installed version
```
├── yeoman-generator@0.21.1
└─┬ yeoman-test@1.0.0
  └── yeoman-generator@0.21.1
```

Log output from pre 0.22.x
```
[19:44:59] Starting 'test'...
****Mon, 11 Jan 2016 00:45:05 GMT yeoman:generator Queueing askFor in prompting
Mon, 11 Jan 2016 00:45:05 GMT yeoman:generator Queueing checkPath in prompting
Mon, 11 Jan 2016 00:45:05 GMT yeoman:generator Queueing copyLocal in configuring
Mon, 11 Jan 2016 00:45:05 GMT yeoman:generator Queueing copyRemote in configuring
Mon, 11 Jan 2016 00:45:05 GMT yeoman:generator Queueing validateSpec in configuring
Mon, 11 Jan 2016 00:45:05 GMT yeoman:generator Queueing server in writing
Mon, 11 Jan 2016 00:45:05 GMT yeoman:generator Queueing genmodels in writing
Mon, 11 Jan 2016 00:45:05 GMT yeoman:generator Queueing genhandlers in writing
Mon, 11 Jan 2016 00:45:05 GMT yeoman:generator Running askFor
GENERATOR-SWAGAPI 2312: prompt results: apiPath => undefined
Mon, 11 Jan 2016 00:45:05 GMT yeoman:generator Running checkPath
GENERATOR-SWAGAPI 2312: apiPath is: undefined
Mon, 11 Jan 2016 00:45:05 GMT yeoman:generator An error occured while running checkPath [Error: missing required input `apiPath`]
.*

6 specs, 0 failures, 5 pending specs
Finished in 3.9 seconds
[19:45:05] Finished 'test' after 6.85 s
```

Installed version
```
├── yeoman-generator@0.22.2
└─┬ yeoman-test@1.0.0
  └── yeoman-generator@0.21.1
```

Log output from 0.22.x
```
[19:41:55] Starting 'test'...
****Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator Queueing askFor in prompting
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator Queueing checkPath in prompting
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator Queueing copyLocal in configuring
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator Queueing copyRemote in configuring
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator Queueing validateSpec in configuring
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator Queueing server in writing
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator Queueing genmodels in writing
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator Queueing genhandlers in writing
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator Running askFor
GENERATOR-SWAGAPI 15420: prompt results: apiPath => undefined
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator Running checkPath
GENERATOR-SWAGAPI 15420: apiPath is: undefined
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator An error occured while running checkPath [Error: missing required input `apiPath`]
.*

6 specs, 0 failures, 5 pending specs
Finished in 1.8 seconds
[19:41:59] Finished 'test' after 3.91 s
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator Running copyLocal
GENERATOR-SWAGAPI 15420: name:  root: C:\cygwin\tmp\ab6bbb449e32437e353b06e4d6c797e1dc3e1af8 project: C:\cygwin\tmp\ab6bbb449e32437e353b06e4d6c797e1dc3e1af8
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator An error occured while running copyLocal [Error: invalid required input `apiPath`]
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator Running copyRemote
GENERATOR-SWAGAPI 15420: apiPath is file: undefined
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator Running validateSpec
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator An error occured while running validateSpec [TypeError: Path must be a string. Received undefined]
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator Running server
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator An error occured while running server [TypeError: Path must be a string. Received undefined]
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator Running genmodels
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator Queueing models in writing
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator Running genhandlers
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator Queueing handlers in writing
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator Queueing tests in writing
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator Running models
Mon, 11 Jan 2016 00:41:59 GMT yeoman:generator An error occured while running models [TypeError: Cannot read property 'definitions' of undefined]
events.js:141
      throw er; // Unhandled 'error' event
      ^

TypeError: Cannot read property 'definitions' of undefined
    at module.exports.yeoman.Base.extend.writing.models (C:\projects\generator-swagapi\models\index.js:42:37)
    at Object.<anonymous> (C:\projects\generator-swagapi\node_modules\yeoman-generator\lib\base.js:436:25)
    at C:\projects\generator-swagapi\node_modules\yeoman-generator\node_modules\run-async\index.js:33:25
    at C:\projects\generator-swagapi\node_modules\yeoman-generator\lib\base.js:446:8
    at processImmediate [as _immediateCallback] (timers.js:383:17)
npm ERR! Test failed.  See above for more details.
```

I've tried to narrow it down but so far have not been able to identify what change(s) would provide such different results. Any feedback would be appreciated.

**Comment by kenjones-cisco** [comment]
The only commit that would seem to have any direct affect would be the following:
https://github.com/yeoman/generator/commit/febdbe289212ae66f572127a9a7ab213c70fad46

**Comment by SBoudrias** [comment]
I'm pretty sure it's always been the case that listening for an error event would not stop the process.

**Comment by SBoudrias** [comment]
Okay, I see the issue. Before when throwing we were not calling `done()`, so the process would stop. Now if it throws, it doesn't stop the loop because it automatically continue.

**Comment by SBoudrias** [comment]
Fixed in `0.22.3`

**Comment by kenjones-cisco** [comment]
@SBoudrias Thanks! :+1:
